libsolidity/parsing/Parser.cpp

```diff
@@ -244,14 +247,20 @@ ASTPointer<ImportDirective> Parser::parseImportDirective()
 	return nodeFactory.createNode<ImportDirective>(path, unitAlias, move(symbolAliases));
 }
 
-ContractDefinition::ContractKind Parser::parseContractKind()
+std::pair<ContractDefinition::ContractKind, bool> Parser::parseContractKind()
 {
 	ContractDefinition::ContractKind kind;
+	bool abstract = false;
 	switch(m_scanner->currentToken())
 	{
 	case Token::Interface:
 		kind = ContractDefinition::ContractKind::Interface;
 		break;
+	case Token::Abstract:
+		abstract = true;
+		m_scanner->next();
+		kind = parseContractKind().first;
```

Comment:
Why do you make a recursive call here? Wouldn't it be simpler to just check the token to be abstract or something else, consume it if it is abstract and then enter the switch?